server/libs/infra/src/repositories/geocoding.repository.ts

```diff
@@ -0,0 +1,44 @@
+import { GeoPoint, ReverseGeocodeResult } from '@app/domain';
+import { localGeocodingConfig } from '@app/infra';
+import { Injectable, Logger } from '@nestjs/common';
+import { getName } from 'i18n-iso-countries';
+import geocoder, { AddressObject, InitOptions } from 'local-reverse-geocoder';
+import { promisify } from 'util';
+
+export interface AdminCode {
+  name: string;
+  asciiName: string;
+  geoNameId: string;
+}
+
+export type GeoData = AddressObject & {
+  admin1Code?: AdminCode | string;
+  admin2Code?: AdminCode | string;
+};
+
+const init = (options: InitOptions): Promise<void> => new Promise<void>((resolve) => geocoder.init(options, resolve));
+const lookup = promisify<GeoPoint[], number, AddressObject[][]>(geocoder.lookUp).bind(geocoder);
+
+@Injectable()
+export class GeocodingRepository {
+  private logger = new Logger(GeocodingRepository.name);
+
+  async init(): Promise<void> {
+    await init(localGeocodingConfig);
+  }
+
+  async reverseGeocode(point: GeoPoint): Promise<ReverseGeocodeResult> {
+    this.logger.debug(`Request: ${point.latitude},${point.longitude}`);
+
+    const [address] = await lookup([point], 1);
+    this.logger.verbose(`Raw: ${JSON.stringify(address, null, 2)}`);
+
+    const { countryCode, name: city, admin1Code, admin2Code } = address[0] as GeoData;
+    const country = getName(countryCode, 'en');
+    const stateParts = [(admin2Code as AdminCode)?.name, (admin1Code as AdminCode)?.name].filter((name) => !!name);
+    const state = stateParts.length > 0 ? stateParts.join(', ') : '';
+    this.logger.debug(`Normalized: ${JSON.stringify({ country, state, city })}`);
+
+    return { country, state, city };
+  }
```

Comment:
State is always empty for me, probably because `admin1Code` and `admin2Code` are a code and there's no lookup happening like there is for country. Might also be better to try and minimize the amount of typecasts, which could make type errors hard to spot especially in the future with package upgrades. 

Also the state column is nullable in the database so why not use null instead of an empty string?

Reply:
I mainly refactored the logic without trying to change how it functioned. It returned empty string before, etc. Happy to make more changes though, especially if there are bugs. 

Whoever implemented it originally had the typecasts. The typing from the library uses `string` for the admin codes, but the response I get back has objects. Do you have some files/gps data that returns strings instead?

Reply:
After recreating my dev instance the state is correctly populated. Before, the lookup function would only return strings for admin codes but I can no longer reproduce it and they are objects. Weird, because I tried it multiple times before. I guess it's solved then and only the typecasts could use some love but no major issues.

Reply:
Looks like it's a problem in the typing for the library. The way it works is they match against the list of cities. The city match contains stuff like country code, state code, etc. They use those codes to create a key that should index into admin1 or admin2 csv file. If they get a match, they return the admin1 or admin2 record. Otherwise they return the original value.

I'm not sure in what situation you would have a record in the cities file that referenced countries/states that did not exist in the admin1 or admin2 files.